### PR TITLE
Fix causal attention mask

### PR DESCRIPTION
**Pull request by thomasw21**
Fixes: #305 

When refactoring some code, I accidentally broke the notion of causal masking. I've fixed it + implemented a test to make sure we don't ever break it again

### CODE REVIEW

File: megatron/model/gpt_model.py [hunk 1]
```diff
@@ -232,13 +232,13 @@ def _to_float16(inputs):
...
         if args.fp32_residual_connection:
-            if hasattr(args, 'attn_mask'):
+            if hasattr(args, 'pretrain_causal_attention') and (args.pretrain_causal_attention is True):
                 self.specs.append(lambda x: x.transpose(0, 1).contiguous().float())
             else:
                 # EmbeddingPipe returns attention mask as well
                 self.specs.append(lambda x: (x[0].transpose(0, 1).contiguous().float(), *x[1:]))
         else:
-            if hasattr(args, 'attn_mask'):
+            if hasattr(args, 'pretrain_causal_attention') and (args.pretrain_causal_attention is True):
```
Comment (stas00) [inline megatron/model/gpt_model.py:241]:
```suggestion
            if hasattr(args, 'pretrain_causal_attention') and args.pretrain_causal_attention:
```

typically one doesn't do `is True`, is there a special purpose for writing it out explicitly? same with all the edits below.

Comment (stas00) [reply megatron/model/gpt_model.py:241]: perhaps this is dealing with 3-state and you don't want to check for `None`?

Comment (stas00) [reply megatron/model/gpt_model.py:241]: but really, get rid of undefined case complete and set the default to be `False`. Problem solved.

Comment (thomasw21) [reply megatron/model/gpt_model.py:241]: Hum okay let's go with `getattr(args, 'pretrain_causal_attention', False)` Thanks!

**Review by stas00 — APPROVED** [review]
Thank you, Thomas!
